core/authenticate/service_test.go

```diff
@@ -290,3 +297,129 @@ func TestService_GetPrincipal(t *testing.T) {
+		Nonce:     "111111", // fixed OTP
+		Metadata: pkgMetadata.Metadata{
+			"callback_url": "",
+		},
+	}
+
+	type args struct {
+		ctx     context.Context
+		request authenticate.RegistrationStartRequest
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *authenticate.RegistrationStartResponse
+		wantErr error
+		setup   func() *authenticate.Service
+	}{
+		{
+			name: "return ErrUnsupportedMethod if request method is not supported",
+			args: args{
+				ctx:     context.Background(),
+				request: authenticate.RegistrationStartRequest{},
+			},
+			want:    nil,
+			wantErr: authenticate.ErrUnsupportedMethod,
+			setup: func() *authenticate.Service {
+				return authenticate.NewService(nil, authenticate.Config{}, nil, nil,
+					nil, nil, nil, nil, nil)
+			},
+		},
+		{
+			name: "simulate a successful StartFlow call",
+			args: args{
+				ctx: context.Background(),
+				request: authenticate.RegistrationStartRequest{
+					Method: authenticate.MailOTPAuthMethod.String(),
+					Email:  "test@example.com",
+				},
+			},
+			want: &authenticate.RegistrationStartResponse{
+				Flow:  flow,
+				State: flow.ID.String(),
+			},
+			wantErr: nil,
+			setup: func() *authenticate.Service {
+				uuid.SetRand(rand.New(rand.NewSource(1)))
+				mockDialer := mailer.NewMockDialer()
+				mockFlowRepo, _, _, _, _ := createMocks(t)
+				ctx := context.Background()
+				_ = strategy.NewMailOTP(mockDialer, "test-subject", "test-body")
+				mockFlowRepo.EXPECT().Set(ctx, flow).Return(nil)
+				srv := authenticate.NewService(
+					nil,
+					authenticate.Config{
+						MailOTP:   authenticate.MailOTPConfig{Validity: 10 * time.Minute},
+						TestUsers: testusers.Config{Enabled: true, OTP: "111111", Domain: "example.com"},
+					},
+					mockFlowRepo, mockDialer, nil, nil,
+					nil, nil, nil)
+				srv.Now = func() time.Time {
+					return timeNow
+				}
+				return srv
+			},
+		},
+		{
+			name: "return sampleErr if flowRepo Set returns error",
+			args: args{
+				ctx: context.Background(),
+				request: authenticate.RegistrationStartRequest{
+					Method: authenticate.MailOTPAuthMethod.String(),
+					Email:  "test@example.com",
+				},
+			},
+			want:    nil,
+			wantErr: sampleErr,
+			setup: func() *authenticate.Service {
+				uuid.SetRand(rand.New(rand.NewSource(1)))
+				mockDialer := mailer.NewMockDialer()
+				mockFlowRepo, _, _, _, _ := createMocks(t)
+				ctx := context.Background()
+				_ = strategy.NewMailOTP(mockDialer, "test-subject", "test-body")
+				mockFlowRepo.EXPECT().Set(ctx, flow).Return(sampleErr)
+				srv := authenticate.NewService(
+					nil,
+					authenticate.Config{
+						MailOTP:   authenticate.MailOTPConfig{Validity: 10 * time.Minute},
+						TestUsers: testusers.Config{Enabled: true, OTP: "111111", Domain: "example.com"},
+					},
+					mockFlowRepo, mockDialer, nil, nil,
+					nil, nil, nil)
+				srv.Now = func() time.Time {
+					return timeNow
+				}
+				return srv
+			},
+		},
+	}
```

Comment:
Can we add a case for SendMail failure as well?

Reply:
Added the failure case for SendMail